libvast/test/fixtures/events.cpp

```diff
@@ -105,96 +180,93 @@ events::events() {
+    for (auto& e : src) {
+      auto bptr = bs.get(e.type());
+      bptr->add(&e);
+      if (bptr->rows() == slice_size)
+        finish_slice(*bptr);
+    }
+    for (auto& [_, builder] : bs.all()) {
+      if (builder.rows() > 0)
+        finish_slice(builder);
+    }
     return slices;
   };
   bro_conn_log_slices = slice_up(bro_conn_log);
-  //bro_dns_log_slices = slice_up(bro_dns_log);
-  //bro_http_log_slices = slice_up(bro_http_log);
-  //bgpdump_txt_slices = slice_up(bgpdump_txt);
+  bro_dns_log_slices = slice_up(bro_dns_log);
+  allocate_id_block(1000); // cause an artificial gap in the ID sequence
+  bro_http_log_slices = slice_up(bro_http_log);
+  bgpdump_txt_slices = slice_up(bgpdump_txt);
   //random_slices = slice_up(random);
   ascending_integers_slices = slice_up(ascending_integers);
   alternating_integers_slices = slice_up(alternating_integers);
   auto to_const_vector = [](const auto& xs) {
     std::vector<const_table_slice_handle> result;
     result.reserve(xs.size());
-    result.insert(result.end(), xs.begin(), xs.end());
+    result.insert(result.begin(), xs.begin(), xs.end());
     return result;
   };
   const_bro_conn_log_slices = to_const_vector(bro_conn_log_slices);
   // const_bro_dns_log_slices = to_const_vector(bro_dns_log_slices);
   // const_bro_http_log_slices = to_const_vector(bro_http_log_slices);
-  // const_bgpdump_txt_slices = to_const_vector(bgpdump_txt_slices);
+  const_bgpdump_txt_slices = to_const_vector(bgpdump_txt_slices);
   // const_random_slices = to_const_vector(random_slices);
   const_ascending_integers_slices = to_const_vector(ascending_integers_slices);
   const_alternating_integers_slices
     = to_const_vector(alternating_integers_slices);
-  auto to_events = [](const auto& slices) {
+  auto sort_by_id = [](std::vector<event>& v) {
+    std::sort(
+      v.begin(), v.end(),
+      [](const auto& lhs, const auto& rhs) { return lhs.id() < rhs.id(); });
+  };
+  auto to_events = [&](const auto& slices) {
     std::vector<event> result;
     for (auto& slice : slices) {
       auto xs = slice->rows_to_events();
       std::move(xs.begin(), xs.end(), std::back_inserter(result));
     }
+    sort_by_id(result);
     return result;
   };
 #define SANITY_CHECK(event_vec, slice_vec)                                     \
   {                                                                            \
     auto flat_log = to_events(slice_vec);                                      \
-    REQUIRE_EQUAL(event_vec.size(), flat_log.size());                          \
-    for (size_t i = 0; i < event_vec.size(); ++i) {                            \
-      if (flatten(event_vec[i]) != flat_log[i]) {                              \
-        FAIL(#event_vec << " != " << #slice_vec);                              \
+    auto sorted_event_vec = event_vec;                                         \
+    sort_by_id(sorted_event_vec);                                              \
```

Comment:
The Ids are grouped by event type, but `event_vec` is still in its original order.

Reply:
Yes, but we assign IDs ascending by group. One vector holds one such group, so how can the vector can end up with non-ascending IDs?

Reply:
I believe `event_vec` holds all events from an input source, in the order they occur in in that input, and mixing event types. 
https://github.com/vast-io/vast/blob/25f0bffb3a1a5a693d8cae12d0a7b51b5f254328/libvast/test/fixtures/events.cpp#L193-L196
Otoh, the ID assignment happens based on event types in the builders, which in turn gets a block of ids from `allocate_id_block()` when a new slice starts.

So for all event sources that contain more than one type of event, the `event_vec` won't be sorted by ID.

For example, if you have a bgpdump with alternating announce and withdraw messages, the IDs in `event_vec` will be in order `[0,100,1,101....].

Reply:
Ah, thanks for the explanation.

@mavam depend some unit tests on the order of these vectors? To me, it would be a nice property if the vectors are sorted by ID. But then, I think we should eventually drop these vectors from the fixture anyways and operate on slices exclusively.

Reply:
>  But then, I think we should eventually drop these vectors from the fixture anyways and operate on slices exclusively.

Right, we won't need those in the future.

> depend some unit tests on the order of these vectors? 

The only problem I can image with respect to the order of these vectors would be in the index, where the IDs map to positions in the bitmaps.

(Some historic context: the idea was that the order of vectors reflects the order how they arrive at the IMPORTER. This is why they have ascending IDs. There's also a gap in the ID sequence to test for non-contiguity in the sequence space.)

Reply:
> The only problem I can image with respect to the order of these vectors would be in the index, where the IDs map to positions in the bitmaps.

Uh, I think we should revise this with our new table slice design. What if we drop all the vectors and instead store *all* events in a `unordered_map<id, event>` for testing purposes? That allows us to fetch events from IDs (sort of like an in-memory ARCHIVE) but otherwise we use the table slices only. What do you guys think?

Btw, I'll go ahead and merge this PR, the discussion seems orthogonal.

Reply:
> What if we drop all the vectors and instead store all events in a `unordered_map<id, event>` for testing purposes? That allows us to fetch events from IDs (sort of like an in-memory ARCHIVE) but otherwise we use the table slices only. What do you guys think?

I think this would fly, in addition to the in table slices for type-specific access (e.g., all Bro conn logs). If we only had the manual archive, it would be cumbersome to get the right subset for writing tests.